lib/enumerable.rb

```diff
@@ -0,0 +1,122 @@
+      my_array[i] = yield to_a[i]
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    to_a.length.times do |i|
+      yield(to_a[i], i)
+    end
+    self
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    my_array = []
+    my_each { |item| my_array << item if yield item }
+    my_array
+  end
+
+  def my_all?(param = nil)
+    if block_given?
+      my_each { |item| return false if yield(item) == false }
+    elsif param.nil?
+      my_each { |item| return false if item.nil? || item == false }
+    elsif !param.nil? && (param.is_a? Class)
+      my_each { |item| return false unless [item.class, item.class.superclass].include?(param) }
+    elsif !param.nil? && param.instance_of?(Regexp)
+      my_each { |item| return false unless param.match(item) }
+    else
+      my_each { |item| return false if item != param }
+    end
+    true
+  end
+
+  def my_any?(param = nil)
+    if block_given?
+      my_each { |item| return true if yield(item) }
+      return false
+    elsif param.nil?
+      my_each { |item| return true if item }
+    elsif !param.nil? && (param.is_a? Class)
+      my_each { |item| return true if [item.class, item.class.superclass].include?(param) }
+    elsif !param.nil? && param.instance_of?(Regexp)
+      my_each { |item| return true if param.match(item) }
+    else
+      my_each { |item| return true if item == param }
+    end
+    false
+  end
+
+  def my_none?(param = nil)
+    if block_given?
+      my_each { |item| return false if yield(item) == true }
+    elsif param.nil?
+      my_each { |item| return false if item == true }
+    elsif !param.nil? && (param.is_a? Class)
+      my_each { |item| return false if [item.class, item.class.superclass].include?(param) }
+    elsif !param.nil? && param.instance_of?(Regexp)
+      my_each { |item| return false if param.match(item) }
+    else
+      my_each { |item| return true unless item == param }
+    end
+    true
+  end
+
+  def my_count(param = nil)
+    count = 0
+    if block_given?
+      my_each { |item| count += 1 if yield item }
+    elsif !param.nil?
+      my_each { |item| count += 1 if item == param }
+    else
+      my_each { |_item| count += 1 }
+    end
+    count
+  end
+
+  def my_map(proc = nil)
+    return to_enum unless block_given? || proc
+
+    my_array = []
+    if proc
+      my_each { |item| my_array << proc.call(item) }
+    else
+      my_each { |item| my_array << yield(item) }
+    end
+    my_array
+  end
+
+  def my_inject(param = nil, operator = nil)
+    raise LocalJumpError if param.nil? && !block_given?
+
+    if block_given?
+      my_each { |item| param = param.nil? ? item : yield(param, item) }
+    elsif operator.nil?
+      operator = param
+      param = nil
+    else
+      operator = operator.to_sym
+      my_each { |item| param = param.nil? ? item : param.send(operator, item) }
+    end
+    param
+  end
+end
```

Comment:
- [ ] when a symbol is specified without an initial value it combines each element of the collection by applying the symbol as a named method

```Ruby
   array = Array.new(100) { rand(0...9) }
   array.my_inject(:+)
```

_expected: 406
got: nil_


- [ ] when a symbol is specified without an initial value it combines each element of the collection by applying the symbol as a named method:range

```Ruby
   range= Range.new(5, 50)
   range.my_inject(:*)
```

_expected: 1267253883404724085150525340252698701849068398706688000000000000
            got: nil_